src/main/java/org/hydev/mcpm/client/database/fetcher/LocalDatabaseFetcher.java

```diff
@@ -0,0 +1,192 @@
+        var localHash = fetchLocalHash();
+
+        // We can't reach the server right now, don't update.
+        if (hostHash == null)
+            return true;
+
+        // Missing hash, we need to update.
+        if (localHash == null)
+            return false;
+
+        return hostHash.equals(localHash);
+    }
+
+    @Nullable
+    private Database fetchLocalDatabase() {
+        try {
+            if (localDatabase != null) {
+                return localDatabase;
+            }
+
+            var file = Paths.get(cacheDirectory.toString(), DATABASE_FILE_NAME).toFile();
+
+            var database = new ObjectMapper().readValue(file, Database.class);
+
+            if (database != null) {
+                localDatabase = database;
+            }
+
+            return database;
+        } catch (IOException e) {
+            return null;
+        }
+    }
+
+    private String readDatabaseFromContent(HttpEntity entity, DatabaseFetcherListener listener) throws IOException {
+        long total = entity.getContentLength();
+
+        var builder = new ByteArrayOutputStream();
+
+        try (var stream = entity.getContent()) {
+            var buffer = new byte[8096];
+            long completed = 0;
+
+            var count = stream.read(buffer);
+
+            while (count > 0) {
+                builder.write(buffer, 0, count);
+
+                completed += count;
+                listener.download(completed, total);
+
+                count = stream.read(buffer);
+            }
+
+            listener.download(total, total);
+        }
+
+        listener.finish();
+
+        return builder.toString(StandardCharsets.UTF_8);
+    }
+
+    @Nullable
+    private Database fetchHostDatabase(DatabaseFetcherListener listener) {
+        try {
+            var response = requestTo(DATABASE_FILE_NAME).execute();
+            var httpResponse = (ClassicHttpResponse) response.returnResponse();
+
+            var body = readDatabaseFromContent(httpResponse.getEntity(), listener);
+
+            var database = new ObjectMapper().readValue(body, Database.class);
+
+            if (database != null) {
+                localDatabase = database;
+
+                try {
+                    var hash = new HashUtils().hash(body);
+
+                    Files.writeString(Paths.get(cacheDirectory.toString(), DATABASE_FILE_NAME), body);
+                    Files.writeString(Paths.get(cacheDirectory.toString(), HASH_FILE_NAME), hash);
+                } catch (NoSuchAlgorithmException e) {
+                    e.printStackTrace();
+                }
+            }
+
+            return database;
+        } catch (IOException e) {
+            return null;
+        }
+    }
+
+    @Override
+    @Nullable
+    public Database fetchDatabase(boolean cache, DatabaseFetcherListener listener) {
+        var databaseHasAlreadyBeenChecked = localDatabase != null;
+
+        if (cache && (databaseHasAlreadyBeenChecked || localDatabaseIsUpToDate())) {
```

Comment:
What if the local database has already been checked but it's out of date? Then it would still return the local version.

Reply:
Yeah, I thought of this but this was intentionally left in this way due to two assumptions:
 - Instances of the CLI will be short lived (we will only check our database once per instance).
 - If we have a long living instance (that handles multiple commands, for example), users can set the `cache` parameter to false to **force** a redownload of the database.

The body of this if statement is expensive, so I want to minimize the need to make multiple subsequent network requests if multiple commands run `fetchDatabase` in sequence.